### PR TITLE
Kestrel diagnostics doc

### PR DESCRIPTION
**Pull request by shirhatti**
**@Rick-Anderson  edit:** [Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/diagnostics?view=aspnetcore-6.0&branch=pr-en-us-23319)

### CODE REVIEW

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+* **Logging** - Structured logs written to [.NET Core logging](xref:fundamentals/logging/index). <xref:Microsoft.Extensions.Logging.ILogger> is used by app frameworks to write logs, and by users for their own logging in an app.
+* **Metrics** - Representation of data measures over intervals of time, for example, requests per second. Metrics are emitted using `EventCounter` and can be observed using [dotnet-counters](/dotnet/core/diagnostics/dotnet-counters) command line tool or with [Application Insights](/azure/azure-monitor/app/eventcounters).
+* **DiagnosticSource** - DiagnosticsSource is a mechanism for production-time logging for rich data payloads for consumption within the process. Unlike logging, which assumes data will leave the process and expects serializable data, DiagnosticSource works well with complex data.
+
+## Logging
+
+Like most components in ASP.NET Core, Kestrel uses `Microsoft.Extensions.Logging` to emit log information. Kestrel employs the use of multiple [categories](xref:fundamentals/logging#log-category-1) which allows you to be selective on which logs you listen to.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/servers/kestrel/diagnostics.md:23]:
```suggestion
Like most components in ASP.NET Core, Kestrel uses `Microsoft.Extensions.Logging` to emit log information. Kestrel employs the use of multiple [categories](xref:fundamentals/logging/index#log-category-1) which allows you to be selective on which logs you listen to.
```

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+| `Microsoft.AspNetCore.Server.Kestrel.BadRequests` | ConnectionBadRequest, RequestProcessingError, RequestBodyMinimumDataRateNotSatisfied |
+| `Microsoft.AspNetCore.Server.Kestrel.Connections` | ConnectionAccepted, ConnectionStart, ConnectionStop, ConnectionPause, ConnectionResume, ConnectionKeepAlive, ConnectionRejected, ConnectionDisconnect, NotAllConnectionsClosedGracefully, NotAllConnectionsAborted, ApplicationAbortedConnection |
+| `Microsoft.AspNetCore.Server.Kestrel.Http2` | Http2ConnectionError, Http2ConnectionClosing, Http2ConnectionClosed, Http2StreamError, Http2StreamResetAbort, HPackDecodingError, HPackEncodingError, Http2FrameReceived, Http2FrameSending, Http2MaxConcurrentStreamsReached |
+| `Microsoft.AspNetCore.Server.Kestrel.Http3` | Http3ConnectionError, Http3ConnectionClosing, Http3ConnectionClosed, Http3StreamAbort, Http3FrameReceived, Http3FrameSending |
+
+### Connection logging
```
Comment (Tratcher) [inline aspnetcore/fundamentals/servers/kestrel/diagnostics.md:33]: Connection logging is already covered here:
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-5.0#connection-logging
Should one of them be removed or abbreviated and linked to the other?

**Review by Tratcher — APPROVED** [review]

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md (hunk 1 above)
Comment (shirhatti) [reply aspnetcore/fundamentals/servers/kestrel/diagnostics.md]: @Rick-Anderson  Does the anchor tag not work on xref links? I want to link to a section of the doc.

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+# Diagnostics in Kestrel
+
+By [Sourabh Shirhatti](https://twitter.com/sshirhatti)
+
+This article provides guidance for gathering diagnostics from Kestrel to help troubleshoot issues. Topics covered include:
+
+* **Logging** - Structured logs written to [.NET Core logging](xref:fundamentals/logging/index). <xref:Microsoft.Extensions.Logging.ILogger> is used by app frameworks to write logs, and by users for their own logging in an app.
+* **Metrics** - Representation of data measures over intervals of time, for example, requests per second. Metrics are emitted using `EventCounter` and can be observed using [dotnet-counters](/dotnet/core/diagnostics/dotnet-counters) command line tool or with [Application Insights](/azure/azure-monitor/app/eventcounters).
+* **DiagnosticSource** - DiagnosticsSource is a mechanism for production-time logging for rich data payloads for consumption within the process. Unlike logging, which assumes data will leave the process and expects serializable data, DiagnosticSource works well with complex data.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/servers/kestrel/diagnostics.md:19]:
```suggestion
* **Logging**: Structured logs written to [.NET Core logging](xref:fundamentals/logging/index). <xref:Microsoft.Extensions.Logging.ILogger> is used by app frameworks to write logs, and by users for their own logging in an app.
* **Metrics**: Representation of data measures over intervals of time, for example, requests per second. Metrics are emitted using `EventCounter` and can be observed using the [dotnet-counters](/dotnet/core/diagnostics/dotnet-counters) command line tool or with [Application Insights](/azure/azure-monitor/app/eventcounters).
* **DiagnosticSource**: `DiagnosticsSource` is a mechanism for production-time logging with rich data payloads for consumption within the process. Unlike logging, which assumes data will leave the process and expects serializable data, `DiagnosticSource` works well with complex data.
```

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md (hunk 2 above)
Comment (shirhatti) [reply aspnetcore/fundamentals/servers/kestrel/diagnostics.md:33]: I'll remove this to link to the existing doc

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+|--|--|
+| `Microsoft.AspNetCore.Server.Kestrel` |  ApplicationError, ConnectionHeadResponseBodyWrite, ApplicationNeverCompleted, RequestBodyStart, RequestBodyDone, RequestBodyNotEntirelyRead, RequestBodyDrainTimedOut, ResponseMinimumDataRateNotSatisfied, InvalidResponseHeaderRemoved, HeartbeatSlow |
+| `Microsoft.AspNetCore.Server.Kestrel.BadRequests` | ConnectionBadRequest, RequestProcessingError, RequestBodyMinimumDataRateNotSatisfied |
+| `Microsoft.AspNetCore.Server.Kestrel.Connections` | ConnectionAccepted, ConnectionStart, ConnectionStop, ConnectionPause, ConnectionResume, ConnectionKeepAlive, ConnectionRejected, ConnectionDisconnect, NotAllConnectionsClosedGracefully, NotAllConnectionsAborted, ApplicationAbortedConnection |
+| `Microsoft.AspNetCore.Server.Kestrel.Http2` | Http2ConnectionError, Http2ConnectionClosing, Http2ConnectionClosed, Http2StreamError, Http2StreamResetAbort, HPackDecodingError, HPackEncodingError, Http2FrameReceived, Http2FrameSending, Http2MaxConcurrentStreamsReached |
+| `Microsoft.AspNetCore.Server.Kestrel.Http3` | Http3ConnectionError, Http3ConnectionClosing, Http3ConnectionClosed, Http3StreamAbort, Http3FrameReceived, Http3FrameSending |
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/servers/kestrel/diagnostics.md:31]:
```suggestion
| `Microsoft.AspNetCore.Server.Kestrel` |  `ApplicationError`, `ConnectionHeadResponseBodyWrite`, `ApplicationNeverCompleted`, `RequestBodyStart`, `RequestBodyDone`, `RequestBodyNotEntirelyRead`, `RequestBodyDrainTimedOut`, `ResponseMinimumDataRateNotSatisfied`, `InvalidResponseHeaderRemoved`, `HeartbeatSlow` |
| `Microsoft.AspNetCore.Server.Kestrel.BadRequests` | `ConnectionBadRequest`, `RequestProcessingError`, `RequestBodyMinimumDataRateNotSatisfied` |
| `Microsoft.AspNetCore.Server.Kestrel.Connections` | `ConnectionAccepted`, `ConnectionStart`, `ConnectionStop`, `ConnectionPause`, `ConnectionResume`, `ConnectionKeepAlive`, `ConnectionRejected`, `ConnectionDisconnect`, `NotAllConnectionsClosedGracefully`, `NotAllConnectionsAborted`, `ApplicationAbortedConnection` |
| `Microsoft.AspNetCore.Server.Kestrel.Http2` | `Http2ConnectionError`, `Http2ConnectionClosing`, `Http2ConnectionClosed`, `Http2StreamError`, `Http2StreamResetAbort`, `HPackDecodingError`, `HPackEncodingError`, `Http2FrameReceived`, `Http2FrameSending`, `Http2MaxConcurrentStreamsReached` |
| `Microsoft.AspNetCore.Server.Kestrel.Http3` | `Http3ConnectionError`, `Http3ConnectionClosing`, `Http3ConnectionClosed`, `Http3StreamAbort`, `Http3FrameReceived`, `Http3FrameSending` |
```

**Comment by Rick-Anderson** [comment]
> @Rick-Anderson Does the anchor tag not work on xref links? I want to link to a section of the doc.

Yes it does. You were missing `/index` on one suggestion I made, but I left the anchor. 

> https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-5.0#connection-logging
>Should one of them be removed or abbreviated and linked to the other?

abbreviate it if you want to keep it.

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+### Connection logging
+
+// TODO: WIP
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/servers/kestrel/diagnostics.md:35]:
```suggestion
<!-- TODO: WIP -->
```

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+var app = builder.Build();
+app.MapGet("/", () => "Hello world");
+app.Run();
+```
+
+## Metrics
+
+Metrics is a representation of data measures over intervals of time, for example, requests per second. Metrics data allows observation of the state of an app at a high-level. Kestrel metrics are emitted using `EventCounter`.
+
+> Unfortunately, the `connections-per-second` and `tls-handshakes-per-second` counters are named incorrectly. Unlike, as implied by them the name, they do not always contain the number of new connections or TLS handshakes per second, but rather the number of new connection or TLS handshakes in the last update interval as requested as the consumer of Events via the `EventCounterIntervalSec` argument in the `filterPayload` to `KestrelEventSource`. It is **recommended** that consumers of these counters scale the metric value based on the `DisplayRateTimeScale` of one second.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/servers/kestrel/diagnostics.md:63]:
```suggestion
> [!NOTE]
> The `connections-per-second` and `tls-handshakes-per-second` counters are named incorrectly. The counters:
>  * Do ***not*** always contain the number of new connections or TLS handshakes per second
>  * Display the number of new connection or TLS handshakes in the last update interval as requested as the consumer of Events via the `EventCounterIntervalSec` argument in the `filterPayload` to `KestrelEventSource`.
>
> We **recommend** consumers of these counters scale the metric value based on the `DisplayRateTimeScale` of one second.
```

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+| Name | Display Name | Description |
+|--|--|--|
+| `connections-per-second` | Connection Rate| The number of new incoming connections per update interval |
+| `total-connections` | Total Connection | The total number of connections |
+| `tls-handshakes-per-second` | TLS Handshake Rate | The number of new TLS handshakes per update interval |
+| `total-tls-handshakes` | Total TLS Handshake | The total number of TLS handshakes |
+| `current-tls-handshakes` | Current TLS Handshakes | The number of TLS handshakes in process |
+| `failed-tls-handshakes` | Failed TLS Handshakes| The total number of failed TLS handshakes |
+| `current-connections` | Current Connections | The total number of connections (including idle connections)
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/servers/kestrel/diagnostics.md:73]:
```suggestion
| `current-connections` | Current Connections | The total number of connections, including idle connections
```

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+| `request-queue-length` | Request Queue Length | The total number requests queued to the thread pool. In a healthy system at steady state, this number should always be close to zero. This metric is unlike the IIS/Http.Sys request queue and cannot be compared  |
+| `current-upgraded-requests` | Current Upgraded Requests (WebSockets) | The number of active WebSocket requests |
+
+## DiagnosticSource
+
+Kestrel emits a `DiagnosticSource` event for HTTP requests rejected at server layer such as malformed requests and protocols violations. As such, these requests never make it into the hosting layer of ASP.NET Core.
+
+Kestrel emits these events with the `Microsoft.AspNetCore.Server.Kestrel.BadRequest` event name and an `IFeatureCollection` as the object payload. The underlying exception can be retrieved by accessing the `IBadRequestExceptionFeature` on the feature collection.
+
+Resolving these events is a two-step process. First, an observer for DiagnosticListener must be created:
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/servers/kestrel/diagnostics.md:84]:
```suggestion
Resolving these events is a two-step process. An observer for `DiagnosticListener` must be created:
```

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+            }
+        }
+    }
+    public void OnError(Exception error) { }
+    public void OnCompleted() { }
+    public virtual void Dispose() => _subscription.Dispose();
+}
+```
+
+Second, you need subscribe to the ASP.NET Core DiagnosticListener with your observer. In our example, we will be creating a callback that logs the underlying exception.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/servers/kestrel/diagnostics.md:120]:
```suggestion
Subscribe to the ASP.NET Core `DiagnosticListener` with the observer. In this example, we create a callback that logs the underlying exception.
```

**Review by Rick-Anderson — APPROVED** [review]

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md (hunk 3 above)
Comment (adityamandaleeka) [inline aspnetcore/fundamentals/servers/kestrel/diagnostics.md:19]: DiagnosticsSource -> DiagnosticSource

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+> Unfortunately, the `connections-per-second` and `tls-handshakes-per-second` counters are named incorrectly. Unlike, as implied by them the name, they do not always contain the number of new connections or TLS handshakes per second, but rather the number of new connection or TLS handshakes in the last update interval as requested as the consumer of Events via the `EventCounterIntervalSec` argument in the `filterPayload` to `KestrelEventSource`. It is **recommended** that consumers of these counters scale the metric value based on the `DisplayRateTimeScale` of one second.
+
+| Name | Display Name | Description |
+|--|--|--|
+| `connections-per-second` | Connection Rate| The number of new incoming connections per update interval |
+| `total-connections` | Total Connection | The total number of connections |
+| `tls-handshakes-per-second` | TLS Handshake Rate | The number of new TLS handshakes per update interval |
+| `total-tls-handshakes` | Total TLS Handshake | The total number of TLS handshakes |
```
Comment (adityamandaleeka) [inline aspnetcore/fundamentals/servers/kestrel/diagnostics.md:70]:
```suggestion
| `total-tls-handshakes` | Total TLS Handshakes | The total number of TLS handshakes |
```

https://github.com/dotnet/aspnetcore/blob/980d7509f869edfa62d2f7aa551c3ff65c258258/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelEventSource.cs#L340

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+## Metrics
+
+Metrics is a representation of data measures over intervals of time, for example, requests per second. Metrics data allows observation of the state of an app at a high-level. Kestrel metrics are emitted using `EventCounter`.
+
+> Unfortunately, the `connections-per-second` and `tls-handshakes-per-second` counters are named incorrectly. Unlike, as implied by them the name, they do not always contain the number of new connections or TLS handshakes per second, but rather the number of new connection or TLS handshakes in the last update interval as requested as the consumer of Events via the `EventCounterIntervalSec` argument in the `filterPayload` to `KestrelEventSource`. It is **recommended** that consumers of these counters scale the metric value based on the `DisplayRateTimeScale` of one second.
+
+| Name | Display Name | Description |
+|--|--|--|
+| `connections-per-second` | Connection Rate| The number of new incoming connections per update interval |
+| `total-connections` | Total Connection | The total number of connections |
```
Comment (adityamandaleeka) [inline aspnetcore/fundamentals/servers/kestrel/diagnostics.md:68]:
```suggestion
| `total-connections` | Total Connections | The total number of connections |
```

https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelEventSource.cs#L315

**Review by adityamandaleeka — APPROVED** [review]

File: aspnetcore/fundamentals/servers/kestrel/diagnostics.md (hunk 6 above)
Comment (shirhatti) [reply aspnetcore/fundamentals/servers/kestrel/diagnostics.md:63]: Oh wow! That's a **LOT** easier to understand

**Comment by shirhatti** [comment]
@Rick-Anderson Good to merge?

**Comment by Rick-Anderson** [comment]
@shirhatti  https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/diagnostics?view=aspnetcore-6.0 
It's live

**Pull request merged**